JDT dependency added in SyncUtil and potentially crashing languages that do not include JDT

I have tested v. 2.38 as part of Eclipse 2025-03 and a dependency to JDT has been introduced in `SyncUtil` (see here: https://github.com/eclipse-xtext/xtext/commit/2f2fe5d60e523ec4b9c5d9a12e96d172833499ea#commitcomment-153274022), which will probably make any DSL not dependent from JDT crash during refactoring operations. It did so for the GAML language (see https://github.com/gama-platform/gama), where JDT has always been treated as an optional dependency. 

Hopefully (thanks @LorenzoBettini)  a workaround can be implemented until this is fixed. It consists in subclassing `SyncUtil` like below and binding this new implementation (`		binder.bind(SyncUtil.class).to(GamlSyncUtil.class);`) in the `UIModule` of your language. 

E.g. 

```java
public class GamlSyncUtil extends SyncUtil {

	/** The workspace. */
	@Inject (optional = true) private IWorkspace workspace;

	/**
	 * Wait for build.
	 *
	 * @param monitor
	 *            the monitor
	 */
	@Override
	public void waitForBuild(final IProgressMonitor monitor) {
		try {
			// waitForJdtIndex(monitor); -- commented out
			workspace.build(IncrementalProjectBuilder.INCREMENTAL_BUILD, monitor);
		} catch (CoreException e) {
			throw new OperationCanceledException(e.getMessage());
		}
	}

}
```

Sure ! Here is the error displayed on the editor when trying to rename an element: 

<img width="362" alt="Image" src="https://github.com/user-attachments/assets/e639bf22-90e1-4bea-a9f4-3db043bed5d3" />

And the corresponding stack trace: 

```
!ENTRY org.apache.log4j 4 0 2025-03-05 08:13:25.594
!MESSAGE org.eclipse.xtext.ui.refactoring.ui.DefaultRenameElementHandler  - Error initializing refactoring

!STACK 0
java.lang.reflect.InvocationTargetException
	at org.eclipse.jface.operation.ModalContext.run(ModalContext.java:396)
	at org.eclipse.jface.dialogs.ProgressMonitorDialog.run(ProgressMonitorDialog.java:470)
	at org.eclipse.ui.internal.progress.ProgressMonitorJobsDialog.run(ProgressMonitorJobsDialog.java:230)
	at org.eclipse.ui.internal.progress.ProgressManager.lambda$26(ProgressManager.java:819)
	at org.eclipse.swt.custom.BusyIndicator.showWhile(BusyIndicator.java:67)
	at org.eclipse.ui.internal.progress.ProgressManager.busyCursorWhile(ProgressManager.java:849)
	at org.eclipse.ui.internal.progress.ProgressManager.busyCursorWhile(ProgressManager.java:828)
	at org.eclipse.ui.internal.progress.ProgressManager.run(ProgressManager.java:977)
	at org.eclipse.xtext.ui.refactoring.ui.SyncUtil.totalSync(SyncUtil.java:61)
	at org.eclipse.xtext.ui.refactoring.ui.SyncUtil.totalSync(SyncUtil.java:54)
	at org.eclipse.xtext.ui.refactoring.ui.DefaultRenameElementHandler.execute(DefaultRenameElementHandler.java:73)
	at org.eclipse.ui.internal.handlers.HandlerProxy.execute(HandlerProxy.java:277)
	at org.eclipse.ui.internal.handlers.E4HandlerProxy.execute(E4HandlerProxy.java:98)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.eclipse.e4.core.internal.di.MethodRequestor.execute(MethodRequestor.java:56)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invokeUsingClass(InjectorImpl.java:299)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invoke(InjectorImpl.java:233)
	at org.eclipse.e4.core.contexts.ContextInjectionFactory.invoke(ContextInjectionFactory.java:174)
	at org.eclipse.e4.core.commands.internal.HandlerServiceHandler.execute(HandlerServiceHandler.java:165)
	at org.eclipse.core.commands.Command.executeWithChecks(Command.java:488)
	at org.eclipse.core.commands.ParameterizedCommand.executeWithChecks(ParameterizedCommand.java:485)
	at org.eclipse.e4.core.commands.internal.HandlerServiceImpl.executeHandler(HandlerServiceImpl.java:204)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.executeCommand(KeyBindingDispatcher.java:308)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.press(KeyBindingDispatcher.java:569)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.processKeyEvent(KeyBindingDispatcher.java:644)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.filterKeySequenceBindings(KeyBindingDispatcher.java:439)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher$KeyDownFilter.handleEvent(KeyBindingDispatcher.java:96)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:91)
	at org.eclipse.swt.widgets.Display.filterEvent(Display.java:1219)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4655)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1657)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1680)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1665)
	at org.eclipse.swt.widgets.Widget.sendKeyEvent(Widget.java:1694)
	at org.eclipse.swt.widgets.Widget.sendKeyEvent(Widget.java:1690)
	at org.eclipse.swt.widgets.Canvas.sendKeyEvent(Canvas.java:522)
	at org.eclipse.swt.widgets.Composite.keyDown(Composite.java:614)
	at org.eclipse.swt.widgets.Display.windowProc(Display.java:6289)
	at org.eclipse.swt.internal.cocoa.OS.objc_msgSendSuper(Native Method)
	at org.eclipse.swt.widgets.Widget.callSuper(Widget.java:239)
	at org.eclipse.swt.widgets.Widget.windowSendEvent(Widget.java:2407)
	at org.eclipse.swt.widgets.Shell.windowSendEvent(Shell.java:2502)
	at org.eclipse.swt.widgets.Display.windowProc(Display.java:6401)
	at org.eclipse.swt.internal.cocoa.OS.objc_msgSendSuper(Native Method)
	at org.eclipse.swt.widgets.Display.applicationSendEvent(Display.java:5705)
	at org.eclipse.swt.widgets.Display.applicationProc(Display.java:5845)
	at org.eclipse.swt.internal.cocoa.OS.objc_msgSend(Native Method)
	at org.eclipse.swt.internal.cocoa.NSApplication.sendEvent(NSApplication.java:117)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4000)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1151)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:339)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1042)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:153)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:668)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:339)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:576)
	at gama.ui.application.Application.start(Application.java:148)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:208)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:149)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:115)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:467)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:298)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:668)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:605)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1481)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1454)
Caused by: java.lang.NoClassDefFoundError: org/eclipse/jdt/internal/core/JavaModelManager
	at org.eclipse.xtext.ui.refactoring.ui.SyncUtil.waitForJdtIndex(SyncUtil.java:161)
	at org.eclipse.xtext.ui.refactoring.ui.SyncUtil.waitForBuild(SyncUtil.java:152)
	at org.eclipse.xtext.ui.refactoring.ui.SyncUtil.doReconcileAndBuild(SyncUtil.java:81)
	at org.eclipse.xtext.ui.refactoring.ui.SyncUtil$1.run(SyncUtil.java:64)
	at org.eclipse.jface.operation.ModalContext$ModalContextThread.run(ModalContext.java:124)
Caused by: java.lang.ClassNotFoundException: org.eclipse.jdt.internal.core.JavaModelManager cannot be found by org.eclipse.xtext.ui_2.38.0.v20250226-0658
	... 5 more
Root exception:
java.lang.NoClassDefFoundError: org/eclipse/jdt/internal/core/JavaModelManager
	at org.eclipse.xtext.ui.refactoring.ui.SyncUtil.waitForJdtIndex(SyncUtil.java:161)
	at org.eclipse.xtext.ui.refactoring.ui.SyncUtil.waitForBuild(SyncUtil.java:152)
	at org.eclipse.xtext.ui.refactoring.ui.SyncUtil.doReconcileAndBuild(SyncUtil.java:81)
	at org.eclipse.xtext.ui.refactoring.ui.SyncUtil$1.run(SyncUtil.java:64)
	at org.eclipse.jface.operation.ModalContext$ModalContextThread.run(ModalContext.java:124)
Caused by: java.lang.ClassNotFoundException: org.eclipse.jdt.internal.core.JavaModelManager cannot be found by org.eclipse.xtext.ui_2.38.0.v20250226-0658
	... 5 more
```

--- a/org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/refactoring/ui/SyncUtil.java
+++ b/org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/refactoring/ui/SyncUtil.java
@@ -158,7 +158,12 @@ public class SyncUtil {
 
 	@SuppressWarnings("restriction")
 	public void waitForJdtIndex(IProgressMonitor monitor) {
-		JavaModelManager.getIndexManager().waitForIndex(true, monitor);
+		try {
+			JavaModelManager.getIndexManager().waitForIndex(true, monitor);
+		} catch (NoClassDefFoundError e) {
+			// ignore it (https://github.com/eclipse-xtext/xtext/issues/3381)
+			// jdt.core is an optional dependency
+		}
 	}
 
 	/**

